[docs] Aggregator architecture readme (#3578)
m3db/m3: site/content/architecture/m3aggregator/flushing.md

---
title: "Flushing"
weight: 2
---

## Overview

Flushing is the process by which m3aggregator instances output the aggregated time series data 
(by using [m3msg](https://github.com/m3db/m3/tree/master/src/msg#readme) protocol).
There are two targets to which the data is flushed:
1. Persistence: the aggregated data is flushed to `m3coordinator` which then 
   persists it in `m3db`.
1. Forwarding: intermediate aggregation data is flushed to other nodes of `m3aggregator`
   (in a multi-node setup) for further processing. This is necessary for rollup rule 
   processing - eg. when a label is dropped, the resulting time series has a new id
   which potentially maps to a different shard (which may be owned by a different node).

> ## Flushing coordination between leaders and followers 

Reviewer: At the bottom of "Flushing coordination between leaders and followers" do you want to also perhaps briefly mention that if/when a follower is promoted to a leader that it will initially perform a flush any time windows that it has buffered but has not been flushed yet (according to the flushed times in etcd).

Author: 👍  https://github.com/m3db/m3/pull/3578/commits/466dc9cc42fe6a32c36850d181e701044c81e24b

for the shardset that they own. The data that has been flushed by the leader node 
is considered as persisted, and the followers can [discard](https://github.com/m3db/m3/blob/0865ebc80e85234b00532f93521438856883da9c/src/aggregator/aggregator/follower_flush_mgr.go#L590)
the corresponding data they have accumulated for the given shard/metrics type/aggregation window.

If, for some reason, a follower node does not receive last flush timestamp updates from
the leader, it will keep the aggregated data for a while, but the follower will go into
[forced flush mode](https://github.com/m3db/m3/blob/0865ebc80e85234b00532f93521438856883da9c/src/aggregator/aggregator/follower_flush_mgr.go#L195-L197) 
after the time period defined by `maxBufferSize` and start discarding the 
accumulated data in order to reclaim the memory. The default value for this setting is 5 minutes:
```yaml
aggregator:
  flushManager:
    maxBufferSize: 5m
```

## Shard cutover/cutoff

[Cutoff/cutover](https://github.com/m3db/m3/blob/0865ebc80e85234b00532f93521438856883da9c/src/cluster/generated/proto/placementpb/placement.proto#L71-L72) 
are timestamps that can be used to control flushing from the given shard.

> They are shard fields in the placement structure.

Reviewer: Perhaps it would read better as:
```suggestion
They are fields that are defined for each shard in the placement structure.
```

If the shard has its cutover field set to some timestamp value, the shard will only
[start flushing](https://github.com/m3db/m3/blob/0865ebc80e85234b00532f93521438856883da9c/src/aggregator/aggregator/list.go#L313) 
once the wall clock will go past the given cutover timestamp. Until this happens, any aggregated
data gets discarded.

Similarly, if the shard has its cutoff field set to some value, the shard will 
[stop flushing](https://github.com/m3db/m3/blob/0865ebc80e85234b00532f93521438856883da9c/src/aggregator/aggregator/list.go#L323-L330) 
once the wall clock will go past the given cutoff timestamp.

> If the shard does not have cutover/cutoff fields set, it is flushing.

Reviewer: Minor suggestion
```suggestion
If the shard does not have cutover/cutoff fields it will flush indefinitely.
```
